### PR TITLE
add(rpc): Add extra `getblock` RPC fields used by some mining pools

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

This PR fixes a confusing error from `s-nomp`, when the block has actually been accepted by Zebra:
> We thought a block was found but it was rejected by the daemon...

### Specifications

https://zcash.github.io/rpc/getblock.html

### Complex Code or Requirements

This RPC makes two separate state requests. But any possible combination of results must be accepted by clients, because they already have to handle chain forks. See the code comments in the PR for details.

## Solution

Changes:
- Add hash, height, and confirmations fields to getblock RPC
    - Some fields are only supplied when it is easy to get them, I added TODOs to provide them all the time
    - This is enough for `s-nomp` for now
- Update snapshots, use new naming scheme
- Remove a test that is already checked by snapshots

Related changes:
- Document the performance requirements of the getblock RPC

Unrelated changes:
- clippy: remove unnecessary return statement

## Review

This isn't on the critical path, but the `s-nomp` error is confusing for users and developers.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow-Up Work

- always provide the hash, height, and confirmations fields for both hash and height arguments
    - make the Depth request support HashOrHeight

### CODE REVIEW

**Comment by teor2345** [comment]
CI panicked in the getblocktemplate-rpcs snapshot code due to a test service timeout, but I didn't change those tests or that code at all:
https://github.com/ZcashFoundation/zebra/actions/runs/4080594241/jobs/7033201856#step:14:3070

It didn't happen in the macOS tests, and I can't reproduce it locally. So I'm going to assume it could be a very slow VM.

Also I got a test failure that is fixed by PR #6091:
https://github.com/ZcashFoundation/zebra/actions/runs/4080594267/jobs/7033830311#step:8:276

**Comment by teor2345** [comment]
Output comparisons:

Height:
```sh
$ zcash-rpc-diff 28232 getblock 1973877 1                                                                                                                                               Checking first node release info...                                                                                                                                                                        Checking second node release info...                                                                                                                                                                       Connected to zebrad (port 28232) and zcashd (zcash-cli zcash.conf port).                                                                                                                                                                                                                                        
Checking zebrad network and tip height...
Checking zcashd network and tip height...
                                                                                                     
Request:                                                                                             
getblock 1973877 1
                                                  
Querying zebrad main chain at height >=1973877...
                                                  
real    0m0.007s                    
user    0m0.004s                      
sys     0m0.003s                
                                                  
Querying zcashd main chain at height >=1973877...
                                                  
real    0m0.007s       
user    0m0.003s         
sys     0m0.003s                     
                                                  
                                                  
Response diff between zebrad and zcashd:
--- /run/user/1000/tmp.n29EjDfk2D.rpc-diff/zebrad-main-1973877-getblock.json    2023-02-06 07:57:59.821727610 +1000
+++ /run/user/1000/tmp.n29EjDfk2D.rpc-diff/zcashd-main-1973877-getblock.json    2023-02-06 07:57:59.828727527 +1000
@@ -1,8 +1,52 @@       
 {                       
+  "hash": "00000000010927b96e49599efaecf1240ddcab9038497b4fcece1be9688309fa",
+  "confirmations": 1,                
+  "size": 2333,                
   "height": 1973877,    
+  "version": 4,                                                                                                                                                                                           
+  "merkleroot": "e9c5f7729846bbb16892dc8d817f391cbe121891536e851ac51957e570837dde",                                                                                                                       
+  "blockcommitments": "fcf47368e1782598534c4f2fb893e2474a4ac355bab3160d08e0ff807f02c1f1",                                                                                                                 
+  "authdataroot": "d4ac15b68ab2a886d18bb2e155bb5e2f8135201ca391a3ea03c61eb833c14fff",                                                                                                                     
+  "finalsaplingroot": "052f17f604954bfceb1b3e106d9c0fda48178edb24026f4ed8f21abf0a18d030",           
+  "finalorchardroot": "3a82f22051ab16b1da85e696fd00cc15fde299dbc717fce917553b3678441b12",
+  "chainhistoryroot": "98b59168f9b179f827681b2b009fd7bfd97c1f037b03163dfbe3fd3db6d62ee1",                                                                                                                 
   "tx": [
     "48297fe8d7e272efb53b5d0947eeb0b94b2f85baa49f47a3b69363012106990c",
     "fe7f0c301a2da277a8e871d3ff4cd1e9cb96986c6472456395ddce8d13d06f0b",
     "d4401d8988c787742357f72d9b70bd7ac982fde8128bb944887d1eaa5f58f832"
-  ]
+  ],
+  "time": 1675634225,
+  "nonce": "b1ad004000000000000000000010000000000000000000000000000000000000",
+  "solution": "...",
+  "bits": "1c014078",
+  "difficulty": 107216921.5625549,
+  "chainwork": "0000000000000000000000000000000000000000000000000c34f2d1790d9bec",
+  "anchor": "558ef286f07e698dd1795c1dd8b157920cda735b50ccc71cf199d55af63bd0c5",
+  "valuePools": [
+    {
+      "id": "sprout",
+      "monitored": true,
+      "chainValue": 26772.46532004,
+      "chainValueZat": 2677246532004,
+      "valueDelta": 0.00000000,
+      "valueDeltaZat": 0
+    },
+    {
+      "id": "sapling",
+      "monitored": true,
+      "chainValue": 998321.04957595,
+      "chainValueZat": 99832104957595,
+      "valueDelta": 0.00000000,
+      "valueDeltaZat": 0
+    },
+    {
+      "id": "orchard",
+      "monitored": true,
+      "chainValue": 34559.95906721,
+      "chainValueZat": 3455995906721,
+      "valueDelta": 0.00000000,
+      "valueDeltaZat": 0
+    }
+  ],
+  "previousblockhash": "0000000000bcdff0b66f5557060d515a7832b923fc2b9502e66d40d999a37075"
 }
```

Hash:
```sh
$ zcash-rpc-diff 28232 getblock 00000000010927b96e49599efaecf1240ddcab9038497b4fcece1be9688309fa 1                                                                          
Checking first node release info...                                                                                                                                                                        
Checking second node release info...                                                                                                                                                                       
Connected to zebrad (port 28232) and zcashd (zcash-cli zcash.conf port).                                                                                                                                   
                                                                                                     
Checking zebrad network and tip height...
Checking zcashd network and tip height...
                                                                                                     
Request:                                                                                             
getblock 00000000010927b96e49599efaecf1240ddcab9038497b4fcece1be9688309fa 1
                                                  
Querying zebrad main chain at height >=1973884...
                                                  
real    0m0.007s                    
user    0m0.005s                      
sys     0m0.002s                
                                                  
Querying zcashd main chain at height >=1973884...
                                                  
real    0m0.007s       
user    0m0.003s         
sys     0m0.003s                     
                                                  
                                                  
Response diff between zebrad and zcashd:
--- /run/user/1000/tmp.DC9batDLY5.rpc-diff/zebrad-main-1973884-getblock.json    2023-02-06 08:10:15.756259394 +1000
+++ /run/user/1000/tmp.DC9batDLY5.rpc-diff/zcashd-main-1973884-getblock.json    2023-02-06 08:10:15.763259315 +1000
@@ -1,9 +1,53 @@       
 {                       
   "hash": "00000000010927b96e49599efaecf1240ddcab9038497b4fcece1be9688309fa",
   "confirmations": 8,                
+  "size": 2333,                
+  "height": 1973877,    
+  "version": 4,
+  "merkleroot": "e9c5f7729846bbb16892dc8d817f391cbe121891536e851ac51957e570837dde",
+  "blockcommitments": "fcf47368e1782598534c4f2fb893e2474a4ac355bab3160d08e0ff807f02c1f1", 
+  "authdataroot": "d4ac15b68ab2a886d18bb2e155bb5e2f8135201ca391a3ea03c61eb833c14fff",
+  "finalsaplingroot": "052f17f604954bfceb1b3e106d9c0fda48178edb24026f4ed8f21abf0a18d030",
+  "finalorchardroot": "3a82f22051ab16b1da85e696fd00cc15fde299dbc717fce917553b3678441b12",
+  "chainhistoryroot": "98b59168f9b179f827681b2b009fd7bfd97c1f037b03163dfbe3fd3db6d62ee1",
   "tx": [                       
     "48297fe8d7e272efb53b5d0947eeb0b94b2f85baa49f47a3b69363012106990c",
     "fe7f0c301a2da277a8e871d3ff4cd1e9cb96986c6472456395ddce8d13d06f0b",
     "d4401d8988c787742357f72d9b70bd7ac982fde8128bb944887d1eaa5f58f832"
-  ]
+  ],
+  "time": 1675634225,
+  "nonce": "b1ad004000000000000000000010000000000000000000000000000000000000",
+  "solution": "...",
+  "bits": "1c014078",
+  "difficulty": 107216921.5625549,
+  "chainwork": "0000000000000000000000000000000000000000000000000c34f2d1790d9bec",
+  "anchor": "558ef286f07e698dd1795c1dd8b157920cda735b50ccc71cf199d55af63bd0c5",
+  "valuePools": [
+    {
+      "id": "sprout",
+      "monitored": true,
+      "chainValue": 26772.46532004,
+      "chainValueZat": 2677246532004,
+      "valueDelta": 0.00000000,
+      "valueDeltaZat": 0
+    },
+    {
+      "id": "sapling",
+      "monitored": true,
+      "chainValue": 998321.04957595,
+      "chainValueZat": 99832104957595,
+      "valueDelta": 0.00000000,
+      "valueDeltaZat": 0
+    },
+    {
+      "id": "orchard",
+      "monitored": true,
+      "chainValue": 34559.95906721,
+      "chainValueZat": 3455995906721,
+      "valueDelta": 0.00000000,
+      "valueDeltaZat": 0
+    }
+  ],
+  "previousblockhash": "0000000000bcdff0b66f5557060d515a7832b923fc2b9502e66d40d999a37075",
+  "nextblockhash": "00000000009cc265164b783c9f0cc50757629a29537d64b5e41a37a716c38dac"
 }
```

The only differences are fields that are not present in Zebra.

File: zebra-rpc/src/methods.rs [hunk 1]
```diff
@@ -578,6 +581,10 @@ where
                     })?;
 
             if verbosity == 0 {
+                // # Peformance
```
Comment (teor2345) [inline zebra-rpc/src/methods.rs]:
```suggestion
                // # Performance
```

File: zebra-rpc/src/methods.rs [hunk 2]
```diff
@@ -601,6 +608,28 @@ where
                     _ => unreachable!("unmatched response to a block request"),
                 }
             } else if verbosity == 1 {
+                // # Peformance
```
Comment (teor2345) [inline zebra-rpc/src/methods.rs]:
```suggestion
                // # Performance
```

File: zebra-rpc/src/methods.rs [hunk 3]
```diff
@@ -601,6 +608,28 @@ where
                     _ => unreachable!("unmatched response to a block request"),
                 }
             } else if verbosity == 1 {
+                // # Performance
+                //
+                // This RPC is used in `lightwalletd`'s initial sync of 2 million blocks,
+                // so it needs to load all its fields very efficiently.
```
Comment (oxarbitrage) [inline zebra-rpc/src/methods.rs:614]: Question: Do lightwalletd uses both verbosity in the initial 2 million blocks ?

File: zebra-rpc/src/methods.rs [hunk 4]
```diff
@@ -601,6 +608,28 @@ where
...
+                //
+                // Currently, we get the transaction IDs from an index, which is much more
+                // efficient than loading all the block data and hashing all the transactions.
+
+                // TODO: look up the hash if we only have a height,
+                //       and look up the height if we only have a hash
+                let hash = hash_or_height.hash().map(GetBlockHash);
+                let height = hash_or_height.height();
+
+                // TODO: do these state  queries in parallel?
```
Comment (oxarbitrage) [inline zebra-rpc/src/methods.rs:624]:
```suggestion
                // TODO: do these state queries in parallel?
```

just an extra space here, not important at all.

**Review by oxarbitrage — APPROVED** [review]
This looks good, i just left a question.

File: zebra-rpc/src/methods.rs (hunk 3 above)
Comment (teor2345) [reply zebra-rpc/src/methods.rs:614]: Yes, unfortunately!

File: zebra-rpc/src/methods.rs (hunk 4 above)
Comment (teor2345) [reply zebra-rpc/src/methods.rs:624]: I'll fix this in another PR so we can unblock mining testing.

File: zebra-rpc/src/methods.rs [hunk 5]
```diff
@@ -611,19 +640,62 @@ where
...
+                        // Confirmations are one more than the depth.
+                        // Depth is limited by height, so it will never overflow an i64.
+                        zebra_state::ReadResponse::Depth(Some(depth)) => Some(i64::from(depth) + 1),
+                        zebra_state::ReadResponse::Depth(None) => {
+                            Some(NOT_IN_BEST_CHAIN_CONFIRMATIONS)
+                        }
+                        _ => unreachable!("unmatched response to a depth request"),
+                    }
+                } else {
+                    // TODO: make Depth support heights as well
```
Comment (arya2) [inline zebra-rpc/src/methods.rs:689]: Could we add this to a follow-up work section and/or create an issue for it?

File: zebra-rpc/src/methods.rs (hunk 4 above)
Comment (arya2) [reply zebra-rpc/src/methods.rs:624]: Related: do we want to open an issue for running the `fetch_mempool_transactions` and `fetch_state_tip_and_local_time` queries in the get_block_template method or any other requests in parallel too?

Comment (teor2345) [reply zebra-rpc/src/methods.rs:624]: We usually leave optional cleanups and minor performance fixes as TODO comments, to keep our issue tracker focused.

If there is a specific user need, we can go ahead and open an issue. But for now I think we're so much faster than `zcashd` here that no-one will ever notice.

File: zebra-rpc/src/methods.rs (hunk 5 above)
Comment (teor2345) [reply zebra-rpc/src/methods.rs:689]: I'll add it to the follow-up work.

Is there a specific RPC client that needs this? That helps us know if we need a ticket, or if a TODO is enough.

Comment (arya2) [reply zebra-rpc/src/methods.rs:689]: > Is there a specific RPC client that needs this?

Not that I know of, a TODO should be enough.

**Pull request merged**